hclsyntax/parser.go

```diff
@@ -1161,15 +1161,19 @@ func (p *parser) finishParsingFunctionCall(name Token) (Expression, hcl.Diagnost
 	for openTok.Type == TokenDoubleColon {
 		nextName := p.Read()
 		if nextName.Type != TokenIdent {
-			diags = append(diags, &hcl.Diagnostic{
+			diag := hcl.Diagnostic{
 				Severity: hcl.DiagError,
 				Summary:  "Missing function name",
 				Detail:   "Function scope resolution symbol :: must be followed by a function name in this scope.",
 				Subject:  &nextName.Range,
 				Context:  hcl.RangeBetween(name.Range, nextName.Range).Ptr(),
-			})
+			}
+			diags = append(diags, &diag)
 			p.recoverOver(TokenOParen)
-			return nil, diags
+			return &ExprSyntaxError{
+				ParseDiags:  hcl.Diagnostics{&diag},
```

Comment:
We typically assume that diagnostics are immutable by convention, despite them being technically mutable, so I don't think it's necessary to make a copy here. It would be highly irregular for something to modify this diagnostic downstream.